### PR TITLE
Support es modules

### PR DESCRIPTION
**Pull request by sigma-andex**
**Description of the change**

This PR adds es  modules support to purescript. It is the rebased pr #3791  from the es modules working group with fixed purity annotations. Solves #3613.

---

**Checklist:**

- [ ] Added a file to CHANGELOG.d for this PR (see CHANGELOG.d/README.md)
- [ ] Added myself to CONTRIBUTORS.md (if this is my first contribution)
- [ ] Linked any existing issues or proposals that this pull request should close
- [ ] Updated or added relevant documentation
- [ ] Added a test for the contribution (if applicable)

### CODE REVIEW

File: src/Language/PureScript/Interactive/IO.hs [hunk 1]
```diff
@@ -21,9 +32,42 @@ onFirstFileMatching f pathVariants = runMaybeT . msum $ map (MaybeT . f) pathVar
...
+    Nothing -> throwError "Could not find node.js version."
+    Just version -> do
+      let semver = do
+            void $ char 'v'
+            major : minor : patch : _ <- fmap (read @Int) (many1 digit) `sepBy` void (char '.')
+            pure (major, minor, patch)
+      case parse (semver <?> "Could not parse node.js version.") "" version of
+        Left err -> throwError $ show err
+        Right (major, minor, patch)
+          | major < 12 -> throwError "Unsupported node.js version."
```
Comment (JordanMartinez) [inline src/Language/PureScript/Interactive/IO.hs:59]: Should we clarify in the error message that at least Node 12 is needed?

File: src/Language/PureScript/Interactive/IO.hs [hunk 2]
```diff
@@ -21,9 +32,42 @@ onFirstFileMatching f pathVariants = runMaybeT . msum $ map (MaybeT . f) pathVar
...
+          | otherwise -> pure (major, minor, patch)
+  let nodeArgs' = if minor < 7 then "--experimental-modules" : nodeArgs else nodeArgs
```
Comment (JordanMartinez) [inline src/Language/PureScript/Interactive/IO.hs:61]: Is this logic correct? The current logic reads that Node 12.5, 13.5, 14.2, 17.5, etc. would all fail this check.

**Comment by thomashoneyman** [comment]
As a note to reviewers, I think the following proposals have support from the core team:

* #4226
* #4227 
* #4230

I believe the plan is to do #4226 and #4230 for PureScript 0.15, but we need `language-javascript` to provide the error message when we detect CommonJS FFI in #4230. We can then do #4227 and remove `language-javascript` in PureScript 0.16.

With that in mind, it's probably OK to only lightly review any bundler changes in this PR, because we are going to remove the bundler in a followup PR.

File: src/Language/PureScript/Interactive/IO.hs (hunk 1 above)
Comment (sigma-andex) [reply src/Language/PureScript/Interactive/IO.hs:59]: makes sense

File: src/Language/PureScript/Interactive/IO.hs (hunk 2 above)
Comment (JordanMartinez) [reply src/Language/PureScript/Interactive/IO.hs]: Resolved by https://github.com/working-group-purescript-es/purescript/pull/11

File: src/Language/PureScript/Interactive/IO.hs [hunk 3]
```diff
@@ -21,9 +32,42 @@ onFirstFileMatching f pathVariants = runMaybeT . msum $ map (MaybeT . f) pathVar
 -- Locates the node executable.
 -- Checks for either @nodejs@ or @node@.
 --
-findNodeProcess :: IO (Maybe String)
-findNodeProcess = onFirstFileMatching findExecutable names
-  where names = ["nodejs", "node"]
+findNodeProcess :: IO (Either String String)
+findNodeProcess = onFirstFileMatching findExecutable ["nodejs", "node"] <&>
+  maybe (throwError "Could not find node.js. Do you have node.js installed and available in your PATH?") pure
```
Comment (rhendric) [inline src/Language/PureScript/Interactive/IO.hs:37]:
```suggestion
  note "Could not find Node.js. Do you have Node.js installed and available in your PATH?"
```

(`note` needs importing from `Protolude`)

File: src/Language/PureScript/Interactive/IO.hs [hunk 4]
```diff
@@ -21,9 +32,42 @@ onFirstFileMatching f pathVariants = runMaybeT . msum $ map (MaybeT . f) pathVar
...
+  result <- readProcessWithExitCode node ["--version"] ""
+  return $ case result of
+    (ExitSuccess, version, _) -> Just version
+    (ExitFailure _, _, _) -> Nothing
+
+readNodeProcessWithExitCode :: Maybe FilePath -> [String] -> String -> IO (Either String (ExitCode, String, String))
+readNodeProcessWithExitCode nodePath nodeArgs stdin = runExceptT $ do
+  process <- maybe (ExceptT findNodeProcess) pure nodePath
+  (major, _, _) <- lift (findNodeVersion process) >>= \case
+    Nothing -> throwError "Could not find node.js version."
```
Comment (rhendric) [inline src/Language/PureScript/Interactive/IO.hs:50]:
```suggestion
    Nothing -> throwError "Could not find Node.js version."
```

File: src/Language/PureScript/Interactive/IO.hs [hunk 5]
```diff
@@ -21,9 +32,42 @@ onFirstFileMatching f pathVariants = runMaybeT . msum $ map (MaybeT . f) pathVar
...
+    Just version -> do
+      let semver = do
+            void $ char 'v'
+            major : minor : patch : _ <- fmap (read @Int) (many1 digit) `sepBy` void (char '.')
+            pure (major, minor, patch)
+      case parse (semver <?> "Could not parse node.js version.") "" version of
```
Comment (rhendric) [inline src/Language/PureScript/Interactive/IO.hs:56]:
```suggestion
      case parse (semver <?> "Could not parse Node.js version.") "" version of
```

File: src/Language/PureScript/Interactive/IO.hs [hunk 6]
```diff
@@ -21,9 +32,42 @@ onFirstFileMatching f pathVariants = runMaybeT . msum $ map (MaybeT . f) pathVar
...
+        Left err -> throwError $ show err
+        Right (major, minor, patch)
+          | major < 12 -> throwError $ "Unsupported node.js version " <> show major <> ". Required node.js version >=12."
```
Comment (rhendric) [inline src/Language/PureScript/Interactive/IO.hs:59]:
```suggestion
          | major < 12 -> throwError $ "Unsupported Node.js version " <> show major <> ". Required Node.js version >=12."
```

**Review by rhendric — COMMENTED** [review]
Not a full review, but since it popped up in my inbox: Node.js is styled as such, not in lowercase.

File: src/Language/PureScript/Errors.hs [hunk 7]
```diff
@@ -701,9 +707,22 @@ prettyPrintSingleError (PPEOptions codeColor full level showDocs relPath) e = fl
       paras [ line $ "In the FFI module for " <> markCode (runModuleName mn) <> ":"
             , indent . paras $
                 [ line $ "The identifier " <> markCode ident <> " contains a prime (" <> markCode "'" <> ")."
-                , line "Primes in identifiers exported from FFI modules are deprecated and won’t be supported in the future."
+                , line "Primes are not allowed in identifiers exported from FFI modules."
```
Comment (JordanMartinez) [inline src/Language/PureScript/Errors.hs:710]: The error type is `DeprecatedFFIPrime`, but this error indicates they are no longer allowed at all. Should a new error be made for them? In other words, if I'm on v0.14.x, I should still be able to go to the doc repo and see the `DeprecatedFFIPrime` page. However, if I'm on v0.15.x in a different project, jumping to the corresponding page shouldn't tell me that these are deprecated. Rather, it should bring me to a separate page that says they are not allowed at all.

File: tests/TestBundle.hs [hunk 8]
```diff
@@ -51,9 +51,8 @@ assertBundles support inputFiles outputFile = do
   case result of
     Left errs -> expectationFailure . P.prettyPrintMultipleErrors P.defaultPPEOptions $ errs
     Right _ -> do
-      process <- findNodeProcess
-      jsFiles <- Glob.globDir1 (Glob.compile "**/*.js") modulesDir
-      let entryPoint = modulesDir </> "index.js"
+      jsFiles <- concat <$> Glob.globDir [Glob.compile "*/*.js", Glob.compile "*/foreign.cjs"] modulesDir
+      let entryPoint = modulesDir </> "index.cjs"
```
Comment (JordanMartinez) [inline tests/TestBundle.hs:55]: Not sure if this matters since it's in the bundler, but I'll comment in case it does matter. Shouldn't this be `index.js`?

File: src/Language/PureScript/Errors.hs (hunk 7 above)
Comment (rhendric) [reply src/Language/PureScript/Errors.hs:710]: Is there no advantage to having one page that says that this is deprecated on <0.15 and not allowed on >=0.15? It might add some weight to the warning for the v0.14-ers—‘X is deprecated’ is a less useful signal than ‘X is deprecated and won't be available in version Y’.

File: app/static/index.js [hunk 9]
```diff
@@ -16,13 +16,24 @@ var evaluate = function evaluate(js) {
...
+    }).replace(/import \* as ([^\s]+) from "([^"]*)"/g, function (_, as, from) {
+        return "var " + as + " = PSCI['" + from.split('/')[1] + "']";
+    }).replace(/export \{([^}]+)\} from "\.\/foreign\.js";?/g, function (_, exports) {
+        return exports.replace(/^\s*([^,\s]+),?\s*$/gm, function (_, exported) {
+            return "module.exports." + exported + " = $foreign." + exported + ";";
+        });
+    }).replace(/export \{([^}]+)\};?/g, function (_, exports) {
+        return exports.replace(/^\s*([^,\s]+)(?: as ([^\s]+))?,?\s*$/gm, function (_, exported, as) {
+            return "module.exports." + (as || exported) + " = " + exported + ";";
+        });
```
Comment (JordanMartinez) [inline app/static/index.js:32]: Should these regexes account for extra spaces? For example
```js
import * as f from "fs"; // normal spacing
import  * as f from "fs"; // typo space
```
The `require` regex on which these are likely based did not need to account for extra whitespace issues.

File: src/Language/PureScript/Errors.hs (hunk 7 above)
Comment (JordanMartinez) [reply src/Language/PureScript/Errors.hs:710]: Good point! I didn't think about this as a migration touch point.

File: app/static/index.js (hunk 9 above)
Comment (sigma-andex) [reply app/static/index.js:32]: afaik this is in the repl in browser feature, which hasn't been updated yet.

File: src/Language/PureScript/Constants/Effect/Unsafe.hs [hunk 10]
```diff
@@ -0,0 +1,9 @@
+module Language.PureScript.Constants.Effect.Unsafe where
+
+import Data.String (IsString)
+
+effectUnsafe :: forall a. (IsString a) => a
+effectUnsafe = "Effect_Unsafe"
+
+unsafePerformEffect :: forall a. (IsString a) => a
+unsafePerformEffect = "unsafePerformEffect"
```
[the remark this answers is not in the document]
Comment (sigma-andex) [reply src/Language/PureScript/Constants/Effect/Unsafe.hs:9]: Seems this was a left over of the old purity annotations. I'll remove it.

**Comment by JordanMartinez** [comment]
@rhendric I'm notifying you on here since you weren't in today's ES module call on Discord. @sigma-andex is fixing various issues we raise in PRs separate from this one. For example:
- `purescript/master` <---- `working-group-purescript-es/purescript:es-modules+purity-annotations` (this PR)
- `working-group-purescript-es/purescript:es-modules+purity-annotations` <-- `fix-pr-comments`  (to address [comments like this](https://github.com/purescript/purescript/pull/4232#discussion_r805016060), we merged working-group-purescript-es/purescript#12)

In the call, I asked how we wanted to complete the ES module support work (e.g. add initial ES support, drop CJS modules, drop `purs bundle`, etc.). For example, should we continue to do what @sigma-andex is doing (i.e. use the `working-group-purescript-es/purescript:es-modules+purity-annotations` repo's branch for doing all ES module work, and resolve new aspects of the ES module work in PRs to that branch), or submit more smaller PRs that implement the functionality, even if some of those PRs don't build (e.g. like this current PR)? @thomashoneyman wasn't on the call due to being on vacation, so I wasn't able to get his input. Ideally, more smaller PRs would be better, but I'm also not sure whether it really matters because the goal will be reached either way. 

Since I was dealing with a move in the past few months, I'm not fully up-to-date on what needs to be done to get this across the finish line, so I'm not sure I'm the best person to talk here. Regardless, I proposed continuing to do what @sigma-andex is doing and implementing everything in 1 PR. 

What say you? 

Also, I proposed making a `v0.14.x` release (`v0.14.6`) before we then focus on merging this PR and getting the `v0.15.x` release underway. Since I know @thomashoneyman wants to get #4233 in, we can still do that by making a separate branch off of `v0.14.6`, adding the work there, and releasing it as `v0.14.7`. As long as he gets that work done before the `v0.15.0` release is made, it shouldn't be a problem. Since he wasn't on the call, I'm going to wait to hear back from him about this. But, I still think a `v0.14.6` release is long overdue.

**Comment by rhendric** [comment]
:+1: to prioritizing a v0.14.6 release.

I don't have any strong feelings about how the ES modules work gets done. Not being a participant in that working group, I only see this stuff when it gets PR'd here. If you have another core member looking at the work as its being done elsewhere and signing off on it at the same quality bar that we'd want for code in the main repo, then I'm fine with whatever happens. If I happen to be the most responsive second core member at the moment, then I guess tell me whether you want me looking at the other repo in real time or if you want me holding off on reviews until there's a PR ready here?

FYI, my current implicit plan was to hold off on doing this big review here until the build was green; I figured your group was going to do a bunch of ecosystem stuff to make that possible. If that's wrong, it'd help to be explicit about what parts you do or don't want me looking at if you're counting on me to be your no. 2.

(And no, I'm not ever on Discord, nor do I plan to be.)

File: app/static/index.js (hunk 9 above)
Comment (JordanMartinez) [reply app/static/index.js:32]: Ah, I see. This code only runs when `purs repl --port somePort` is called. I've always just used `purs repl`, which is always using the Node backend. 

> which hasn't been updated yet.

But shouldn't this be fixed here? Or does that mess up tests and hence why it hasn't yet?

**Comment by JordanMartinez** [comment]
So, my working theory for why CI is failing is because `bower` is pulling in transitive deps that don't refer to the `es-modules` branch. For example, in `tests/support/bower.json`, `purescript-console` refers to the `es-modules` branch. However, `psci-support` refers to `v5.0.0`, and `v5.0.0`'s `bower.json` file refers to `purescript-console` at `v5.0.0`.

Making CI pass might entail updating all repos' `bower.json` files to reference the es-working-group repo's `es-modules` branch for the corresponding repos. For now, I'm just updating the deps of repos in `tests/support/bower.json` that don't refer to an `es-modules` branch.

**Comment by JordanMartinez** [comment]
@sigma-andex Per Nate's guidance, I added the resolutions I got after running `bower install` manually and resolving to use the `es-modules` version of packages when a conflict occurred. I've pushed that commit here.

The Windows build is still building as I write this, but CI otherwise passes. The Linux build fails because it's attempting to compile the package set. To disable that, you'll need to add a file whose name starts with `break` in the `CHANGELOG.d/` directory. Then, that step will be skipped and the build should be green. I would do that myself, but I feel like you and Mark are the ones better equipped to write the changelog entry for that.

**Comment by JordanMartinez** [comment]
CI failed because the script that prevents the package set from building expects the file name to start with `breaking`, not `break` as I instructed. Sorry about that. I've fixed the issue, so CI should build after these changes.

File: app/static/index.js (hunk 9 above)
Comment (natefaubion) [reply app/static/index.js:32]: It's not clear to me what the resolution was here. Do we need to note this in the release? Will it be updated or will it be removed? Will that happen in a subsequent PR?

File: src/Language/PureScript/CodeGen/JS/Printer.hs [hunk 11]
```diff
@@ -119,6 +120,30 @@ literals = mkPattern' match'
...
+    , return $ emit "\n"
+    , currentIndent
+    , return . emit $ "}" <> maybe "" ((" from " <>) . prettyPrintStringJS) from
+    ]
+    where
+    exportedIdentToString Nothing ident
+      | nameIsJsReserved ident || nameIsJsBuiltIn ident
+      = "$$" <> ident <> " as " <> ident
+    exportedIdentToString _ "$main"
+      = T.concatMap identCharToText "$main" <> " as $main"
```
Comment (natefaubion) [inline src/Language/PureScript/CodeGen/JS/Printer.hs:144]: I'm not totally sure what this is about? Isn't this an identity?

File: src/Language/PureScript/Errors.hs [hunk 12]
```diff
@@ -701,9 +707,22 @@ prettyPrintSingleError (PPEOptions codeColor full level showDocs relPath) e = fl
...
+      paras [ line $ "A CommonJS foreign module implementation was provided for module " <> markCode (runModuleName mn) <> ": "
+            , indent . lineS $ path
+            , line "CommonJS foreign modules are deprecated and won't be supported in the future."
+            ]
+    renderSimpleErrorMessage (UnsupportedFFICommonJSExports mn idents) =
+      paras [ line $ "The following CommonJS exports are not supported in the ES foreign module for module " <> markCode (runModuleName mn) <> ": "
+            , indent . paras $ map line idents
+            ]
+    renderSimpleErrorMessage (UnsupportedFFICommonJSImports mn mids) =
+      paras [ line $ "The following CommonJS imports are no supported in the ES foreign module for module " <> markCode (runModuleName mn) <> ": "
```
Comment (natefaubion) [inline src/Language/PureScript/Errors.hs:723]:
```suggestion
      paras [ line $ "The following CommonJS imports are not supported in the ES foreign module for module " <> markCode (runModuleName mn) <> ": "
```

File: app/static/index.js (hunk 9 above)
Comment (sigma-andex) [reply app/static/index.js:32]: Wasn't the idea to remove the browser repl in a subsequent PR?

Comment (natefaubion) [reply app/static/index.js:32]: I'm fine with removing it, but it wasn't clear to me there was any consensus here with the core team. Did I miss that?

**Review by natefaubion — APPROVED** [review]

File: app/static/index.js (hunk 9 above)
Comment (JordanMartinez) [reply app/static/index.js:32]: I've missed a few meetings, but I don't think there has been consensus yet as I don't think this has come up. I'm up for removing it though because I've never used the browser version. If I want a REPL-like experience, I usually just use Try PureScript because there's a small bug in one part of my code and starting a REPL just to debug that one function isn't worth it to me.

**Comment by JordanMartinez** [comment]
Just a reminder. This PR should not be merged until after #4236 gets merged and released. Once that's done, we're ready to start the v0.15.x migration work.

**Comment by sigma-andex** [comment]
Can someone approve the workflow to build again. tests are fixed now

**Comment by JordanMartinez** [comment]
@rhendric CI builds. Could you give this a review, too?

File: CHANGELOG.d/breaking_switch-to-es-modules.md [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+  ...or using the short version...
+
+  ```javascript
+  export { myvar } from 'mymodule';
+  ```
+
+* FFI is annotated with `/* #__PURE__ */` so that bundlers can perform DCE
+* If CJS is detected a `Warning` is emitted
+* The current LTS Node.js version `12` is now the required minimum version
+* `purs bundle` has been rudimentarily updated but will be removed in a subsequent PR
```
Comment (thomashoneyman) [inline CHANGELOG.d/breaking_switch-to-es-modules.md:28]: Does this line need to be included? After all, once we remove the bundler the combined changelog will reflect both of these changes.

File: CHANGELOG.d/breaking_switch-to-es-modules.md [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+  export const myvar = M.myvar
+  ```
+  ...or using the short version...
+
+  ```javascript
+  export { myvar } from 'mymodule';
+  ```
+
+* FFI is annotated with `/* #__PURE__ */` so that bundlers can perform DCE
+* If CJS is detected a `Warning` is emitted
```
Comment (thomashoneyman) [inline CHANGELOG.d/breaking_switch-to-es-modules.md:26]: Relatedly, should we include this line, given that in a followup we are going to make this an error? I know that this is true in _this_ pull request, but we have immediate plans to change it.

File: CHANGELOG.d/breaking_switch-to-es-modules.md (hunk 13 above)
Comment (JordanMartinez) [reply CHANGELOG.d/breaking_switch-to-es-modules.md]: Oh, I was thinking that line would be removed in the PR that removes it. Each of these changelog entry files can be edited multiple times and each PR that edits it will be included in the list of PRs for that entry.

Comment (thomashoneyman) [reply CHANGELOG.d/breaking_switch-to-es-modules.md]: Ah, I see. In that case carry on :)

File: src/Language/PureScript/CodeGen/JS/Printer.hs (hunk 11 above)
Comment (JordanMartinez) [reply src/Language/PureScript/CodeGen/JS/Printer.hs:144]: @sigma-andex Any comments on this?

Comment (sigma-andex) [reply src/Language/PureScript/CodeGen/JS/Printer.hs:144]: Apparently `$main` is generated by the REPL (PSCI volatile module) and needs to be aliased. I just tested and without it the repl doesnt work. As far as I understood the $main is exported as $dollarmain (due to identCharToText) and because of this needs to be aliased.

Comment (natefaubion) [reply src/Language/PureScript/CodeGen/JS/Printer.hs:144]: This is likely due to the renamer here being severely outdated. Dollars are not valid in PS names. At one point, we supported operators without named aliases, and so it was important to rename `$` to `$dollar`. This obviously has not been the case for _some_ time, but that code was clearly never changed. We can leave this for now.

File: src/Language/PureScript/CodeGen/JS.hs [hunk 15]
```diff
@@ -65,40 +66,44 @@ moduleToJs (Module _ coms mn _ imps exps reExps foreigns decls) foreign_ =
...
+  annotatePure js@(AST.ObjectLiteral _ []) = js
+  annotatePure js@(AST.ObjectLiteral _ _) = pureIife js
+  annotatePure js@(AST.ArrayLiteral _ []) = js
+  annotatePure js@(AST.ArrayLiteral _ _) = pureIife js
+  annotatePure (AST.VariableIntroduction ss name (Just js)) = AST.VariableIntroduction ss name (Just (annotatePure js))
+  annotatePure (AST.Comment a b js) = AST.Comment a b (annotatePure js)
+  annotatePure js = js
+
+  pureIife :: AST -> AST
+  pureIife val = AST.Pure Nothing $ AST.App Nothing (AST.Function Nothing Nothing [] (AST.Block Nothing [AST.Return Nothing val])) []
```
Comment (rhendric) [inline src/Language/PureScript/CodeGen/JS.hs:100]: How was this arrived at? In particular, it seems a little heavy-handed to wrap every top-level non-empty object or array in an IIFE, and I think for esbuild the only things that need this annotation are function calls not in the scope of a function body. Would a more targeted approach not have worked/was it tried and found wanting/is there another bundler the working group was looking at that has different needs?

File: src/Language/PureScript/CoreImp/AST.hs [hunk 16]
```diff
@@ -92,6 +93,12 @@ data AST
   -- ^ instanceof check
   | Comment (Maybe SourceSpan) [Comment] AST
   -- ^ Commented JavaScript
+  | Pure (Maybe SourceSpan) AST
```
Comment (rhendric) [inline src/Language/PureScript/CoreImp/AST.hs:96]: Why add `Pure` to the CoreImp AST, instead of using a `Comment` node? We never match on it except to emit JavaScript, and it'll be one more constructor to have to handle when writing CoreImp optimization passes. Are there non-JS backends that share our CoreImp AST instead of working from CoreFn or something?

File: tests/TestBundle.hs (hunk 8 above)
Comment (rhendric) [reply tests/TestBundle.hs:55]: IIUC, this imperfectly-named variable is the *output* of the bundler, which is a CommonJS module, and we seem to be sticking to the convention that `.js` means ES module and `.cjs` means CommonJS module.

**Review by rhendric — COMMENTED** [review]
Looks pretty sane on first pass, although I have some implementation decision questions (that may just net out to me not fully understanding the requirements).

File: purescript.cabal [hunk 17]
```diff
@@ -2,7 +2,7 @@ cabal-version:  2.4
 
 name:           purescript
 -- note: When updating the prerelease identifier, update it in app/Version.hs too!
-version:        0.14.5
+version:        0.15.0
```
[the remark this answers is not in the document]
Comment (sigma-andex) [reply purescript.cabal:5]: We changed this because it made testing much easier so we could see that the right compiler and package sets were used. Is there any advantage in removing it now and adding it back in a subsequent PR/commit?

[the remark this answers is not in the document]
Comment (rhendric) [reply purescript.cabal:5]: Let me rephrase: will the build break if this is not merged with the PR? We have a whole release process during which version numbers are updated; it's not like you're responsible for bumping the version number just because you're making a breaking change. But if there's something about the parallel package set you're using that requires this for CI to pass, then we should know about that.

[the remark this answers is not in the document]
Comment (sigma-andex) [reply purescript.cabal:5]: The parallel package sets  [`metadata.version` to `v0.15`](https://github.com/working-group-purescript-es/package-sets/blob/main/packages.dhall#L111). But if I'm not mistaken, we are not using spago package-sets here in the compiler repo, but rather bower. So I think it will not break, the only thing is that the master branch will not work with the package-set from the working group until the release increases the version.

File: src/Language/PureScript/CodeGen/JS.hs [hunk 18]
```diff
@@ -197,10 +213,13 @@ moduleToJs (Module _ coms mn _ imps exps reExps foreigns decls) foreign_ =
...
-  accessor (Ident prop) = accessorString $ mkString prop
-  accessor (GenIdent _ _) = internalError "GenIdent in accessor"
-  accessor UnusedIdent = internalError "UnusedIdent in accessor"
+  moduleAccessor :: Ident -> AST -> AST
+  moduleAccessor (Ident prop) = moduleAccessorString prop
+  moduleAccessor (GenIdent _ _) = internalError "GenIdent in moduleAccessor"
+  moduleAccessor UnusedIdent = internalError "UnusedIdent in moduleAccessor"
+
+  moduleAccessorString :: Text -> AST -> AST
+  moduleAccessorString = accessorString . mkString . T.replace "'" "$prime"
```
[the remark this answers is not in the document]
Comment (sigma-andex) [reply src/Language/PureScript/CodeGen/JS.hs:222]: I think @natefaubion is more qualified to answer theses questions.

File: purescript.cabal (hunk 17 above)
[the remark this answers is not in the document]
Comment (sigma-andex) [reply purescript.cabal:5]: Will remove this then.

[the remark this answers is not in the document]
Comment (sigma-andex) [reply purescript.cabal]: Done

File: src/Language/PureScript/CodeGen/JS.hs (hunk 18 above)
[the remark this answers is not in the document]
Comment (natefaubion) [reply src/Language/PureScript/CodeGen/JS.hs:222]: I don't think I wrote this particular bit, but I think this is here because it's a tweak/extension of code that was already here which worked for CJS but needed to change for ESM. I don't see why we couldn't try to put it with other escaping code, I just don't know what the exact context is.

File: src/Language/PureScript/CodeGen/JS.hs (hunk 15 above)
Comment (natefaubion) [reply src/Language/PureScript/CodeGen/JS.hs:100]: For specific numbers, with a targeted approach, it was difficult to get esbuild to even beat purs-bundle, despite using a much lighter weight bundle encoding. It was leaving in tons of unused, seemingly trivial things, which is how I arrived at the record accessor realization. With this change, esbuild beats purs-bundle by about 30%.

**Comment by JordanMartinez** [comment]
CI is failing because I deleted the `es-modules` branch for the `purescript-lists` repo since it doesn't have any FFI. I'll push a commit that reverts it back to `6.0.0` after verifying that the resolutions field doesn't need to also be updated.

File: src/Language/PureScript/CoreImp/AST.hs (hunk 16 above)
Comment (natefaubion) [reply src/Language/PureScript/CoreImp/AST.hs:96]: I think at one point it wrapped a full expression with parens and all that. Now that it just inserts a comment we can probably remove it.

Comment (natefaubion) [reply src/Language/PureScript/CoreImp/AST.hs:96]: > Are there non-JS backends that share our CoreImp AST instead of working from CoreFn or something?

I don't think there are.

File: src/Language/PureScript/CodeGen/JS.hs (hunk 18 above)
[the remark this answers is not in the document]
Comment (rhendric) [reply src/Language/PureScript/CodeGen/JS.hs:222]: It seems to me that this is necessary only because we switched from emitting `exports = { "foo'": foo$prime }` to `export { foo$prime }` and thus need to reference it with `Module_Name["foo$prime"]`, instead of `Module_Name["foo'"]` as we do currently. If we instead emit `export { foo$prime as "foo'" }` then this wouldn't have needed to change, I think. So it's either add mangling code here or modify the existing mangling code in `JS.Printer`, of which I think the latter is preferable. Decent chance I've gotten a detail wrong here though.

File: src/Language/PureScript/CodeGen/JS.hs (hunk 15 above)
Comment (rhendric) [reply src/Language/PureScript/CodeGen/JS.hs:100]: That's disappointing, but very believable. I'm fine with moving forward with this as-is, but do you by chance have a script for running those benchmarks? I'd be interested in seeing if I can improve this at some point; extra IIFEs are not as simple to strip out as comments are.

Comment (natefaubion) [reply src/Language/PureScript/CodeGen/JS.hs:100]: You can follow my journey and implementations in the original PR to the working group https://github.com/working-group-purescript-es/purescript/pull/7

File: src/Language/PureScript/CodeGen/JS.hs (hunk 18 above)
[the remark this answers is not in the document]
Comment (natefaubion) [reply src/Language/PureScript/CodeGen/JS.hs:222]: > export { foo$prime as "foo'" }

I wasn't aware this was valid syntax. Wild!

[the remark this answers is not in the document]
Comment (rhendric) [reply src/Language/PureScript/CodeGen/JS.hs:222]: 🤦 That's because it isn't, at least per the spec. It works in Node and I didn't bother to check beyond that. Mea culpa.

This is all fine then.

[the remark this answers is not in the document]
Comment (natefaubion) [reply src/Language/PureScript/CodeGen/JS.hs:222]: It is according to the [most recent spec](https://tc39.es/ecma262/#prod-ModuleExportName) at least. But I don't think we can rely on the most recent spec.

File: src/Language/PureScript/CodeGen/JS.hs (hunk 15 above)
Comment (natefaubion) [reply src/Language/PureScript/CodeGen/JS.hs:100]: As far as the scripts, they were in a branch of real-world-halogen I think. I don't remember exactly at this point. @sigma-andex might remember.

Comment (sigma-andex) [reply src/Language/PureScript/CodeGen/JS.hs:100]: here is the link: https://github.com/working-group-purescript-es/challenges/tree/main/bundle-challenge#purescript-halogen-template

Comment (rhendric) [reply src/Language/PureScript/CodeGen/JS.hs:100]: Thank you both!

**Comment by JordanMartinez** [comment]
For those not watching the `working-group-purescript-es/purescript` repo, @rhendric opened a PR addressing his [concern about the extra IIFEs](https://github.com/purescript/purescript/pull/4232/#discussion_r813557840) and his [concern about the `Pure` constructor being separate from the `Comment` constructor](https://github.com/purescript/purescript/pull/4232/#discussion_r813559583), which hinders optimization passes later, in https://github.com/working-group-purescript-es/purescript/pull/15

[His comment about the `.cjs` file extension](https://github.com/purescript/purescript/pull/4232/#discussion_r813561483) was not addressed in the above PR.

**Comment by JordanMartinez** [comment]
> IIUC, this imperfectly-named variable is the output of the bundler, which is a CommonJS module, and we seem to be sticking to the convention that .js means ES module and .cjs means CommonJS module.

Since we'll be dropping `purs bundle`, I don't think this issue needs to be dealt with.

**Review by thomashoneyman — APPROVED** [review]

**Comment by rhendric** [comment]
> Since we'll be dropping `purs bundle`, I don't think this issue needs to be dealt with.

I was possibly unclear; I was justifying the change, not saying that it's an issue that needs to be resolved. Nets out to the same result as your conclusion though. 😄

**Review by rhendric — APPROVED** [review]
LGTM (naturally, now that I've gotten my hands into it), but what's going on with [this thread about the browser REPL backend](https://github.com/purescript/purescript/pull/4232/files#r805061844)?

**Comment by natefaubion** [comment]
@rhendric The browser REPL is effectively broken because it relies on purs bundle. Since we are removing purs bundle in a follow up we need to either decide:
* Fix up it's static server to serve individual modules.
* Remove the browser REPL altogether.

I'm personally in favor of the second option, because I'm not aware of anyone taking advantage of it, and I find the experience of try-purescript to be so much better for testing things in the browser (try-purescript will need to be updated at some point too). In our discussions, no one expressed interest in maintaining this feature. However a consensus will need to be reached among the core team as a whole. If others on the team are interested in using and maintaining it, that's fine.

**Comment by rhendric** [comment]
I've never used this feature either (and would personally be fine with removing it if due diligence is done), but my understanding is that it's not a REPL in the browser; it's a browser-hosted backend for the console REPL. In a sense it's the inverse of Try PureScript, which is a server-hosted backend with a browser frontend. I think it's potentially uniquely useful for people doing frontend development in PureScript and who want to test some foreign JS integration not supported in Node.js.

**Comment by natefaubion** [comment]
> I've never used this feature either (and would personally be fine with removing it if due diligence is done), but my understanding is that it's not a REPL in the browser; it's a browser-hosted backend for the console REPL. In a sense it's the inverse of Try PureScript, which is a server-hosted backend with a browser frontend. I think it's potentially uniquely useful for people doing frontend development in PureScript and who want to test some foreign JS integration not supported in Node.js.

The browser repl just pushes evaluation to the browser. In that sense, it's not different from try purescript, which also evaluates in the browser, except you also have an editor right there as well.

**Comment by rhendric** [comment]
Oh, ignore me then!

**Comment by JordanMartinez** [comment]
`v0.14.7` has been released.

**Comment by JordanMartinez** [comment]
I merged the following PRs in the following order to see what merge conflicts might arise. All merge conflicts are minor:
- #4149 
- #4243
- #4241 
- #4240 
- #4239 

I'll be merging these PRs in the above order once CI passes since all of them are already approved.

**Comment by rhendric** [comment]
#4241 is approved?

**Comment by JordanMartinez** [comment]
Oh, it's not yet. Sorry about that, and thanks for pointing that out.

**Comment by JordanMartinez** [comment]
I'm trying to get #4241 in. Then I'd like to merge #4248. After that, yeah.

**Comment by JordanMartinez** [comment]
> I'm trying to get https://github.com/purescript/purescript/pull/4241 in. Then I'd like to merge https://github.com/purescript/purescript/pull/4248. After that, yeah.

Since #4241 will need a more in-depth review and may possibly require us to merge a PR dropping `purs bundle` before it'll build, I'd like to merge #4249 (can I get an approval once CI builds?) to fix CI, and then merge #4248. After that, we'll be done for this round of PRs, I think.
